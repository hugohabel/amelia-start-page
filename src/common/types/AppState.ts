// App State
export interface AppState {
  widgets: Widget[];
  sidebar: Sidebar;
}

// Sidebar
export interface Sidebar {
  isOpen: boolean;
  activeMenu: ActiveMenu;
  activeView: ActiveView;
}

// Widget
interface Widget {
  type: WidgetType;
  enabled: boolean;
  data: CountdownWidget;
}

// Countdown Widget
interface CountdownWidget {
  events: Event[];
}

interface Event {
  id: string;
  name: string;
  date: string;
  background: string;
}
// End Countdown Widget

// Auxiliary Types
type WidgetType = 'countdown';
export type ActiveMenu = 'add' | 'about' | 'config' | null;
export type ActiveView = 'add_countdown' | null;

// App State Actions
type Action =
  | { type: 'addNewWidget'; widget: Widget }
  | { type: 'toggleSidebar'; }
  | { type: 'setActiveMenu'; activeMenu: ActiveMenu }
  | { type: 'setActiveView'; activeView: ActiveView }
  | { type: 'addNewCountdownEvent'; event: Event }
  | { type: 'removeWidgetItem'; uuid: string; widgetType: string };

export type AppAction = Action;
export type AppDispatch = (action: AppAction) => void;
