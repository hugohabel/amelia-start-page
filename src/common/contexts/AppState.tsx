// External Dependencies
import { createContext, ReactNode, useContext, useReducer } from 'react';
import { debounce } from 'lodash';

// Internal Dependencies
import { LOCAL_STORAGE_STATE_KEY } from '../constants';
import { TAppAction, TAppDispatch, IAppState } from '../types/AppState';
import { SafeLocalStorage } from '../utils/localStorage/localStorage';
import { initialAppState, rootReducer } from '../state/state';

// Interfaces + Types
interface IAppStateProviderProps {
  children: ReactNode;
  initialStateOverride?: IAppState;
}

// Utility Functions

/**
 * Retrieves state from local storage or initializes with default state.
 *
 * @param defaultState AppState   Default app state.
 * @returns AppState  State (can be an existing one, or a default one)
 */
function getStateFromLocalStorageOrInitialize(defaultState: IAppState) {
  const stateFromLocalStorage = SafeLocalStorage.getInstance().getItem(LOCAL_STORAGE_STATE_KEY);
  if (stateFromLocalStorage) {
    return JSON.parse(stateFromLocalStorage) as IAppState;
  }
  return defaultState;
}

/**
 * Persists app state in memory or in local storage
 *
 * @param state AppState  State to be persisted.
 */
function persistState(state: IAppState) {
  SafeLocalStorage.getInstance().setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: false,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: IAppState, action: TAppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

// Create Contexts
const AppStateContext = createContext<IAppState>(initialAppState);
const AppDispatchContext = createContext<TAppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn('Calling default dispatch is a no-op!')
);

/**
 * App State Provider
 *
 * @param param AppStateProviderProps Props to initialize the app state provider.
 * @returns JSX App Dispatch and App State Contexts.
 */
export function AppStateProvider({ children, initialStateOverride }: IAppStateProviderProps) {
  const hydratedInitialAppState = {
    ...initialAppState,
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    getStateFromLocalStorageOrInitialize
  );

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
}

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);
