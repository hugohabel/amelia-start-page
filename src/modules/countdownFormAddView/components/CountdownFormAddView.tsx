// External Dependencies
import ArrowBackOutlinedIcon from '@material-ui/icons/ArrowBackOutlined';

// Internal Dependencies
import { useAppDispatch } from '../../../common/contexts/AppState';
import CountdownForm from '../../countdownForm/components/CountdownForm';

// Styles
import styles from './CountdownFormAddView.module.css';

function CountdownFormAddView() {
  const dispatch = useAppDispatch();

  // Handler Functions
  function handleGoBackClick() {
    dispatch({
      type: 'setActiveView',
      activeView: null,
    });
  }

  return (
    <>
      {/* Go Back */}
      <div data-testid="go-back" className={styles.backContainer} onClick={handleGoBackClick}>
        <ArrowBackOutlinedIcon fontSize="small" />
        <span>Go Back</span>
      </div>
      {/* End Go Back */}

      {/* Title */}
      <h2 className={styles.title}>New countdown</h2>
      <p className={styles.subtitle}>Add the details for your countdown widget</p>
      {/* End Title */}

      {/* Countdown Form */}
      <CountdownForm />
      {/* End Countdown Form */}
    </>
  );
}

export default CountdownFormAddView;
