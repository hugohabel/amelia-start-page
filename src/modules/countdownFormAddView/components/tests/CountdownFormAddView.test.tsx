// External Dependencies
import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

// Internal Dependencies
import { AppStateProvider, useAppDispatch } from '../../../../common/contexts/AppState';
import CountdownFormAddView from '../CountdownFormAddView';

// Mocked Functions
jest.mock('../../../../common/contexts/AppState', () => ({
  ...(jest.requireActual('../../../../common/contexts/AppState') as object),
  useAppDispatch: jest.fn(),
}));
let mockAppDispatch: ReturnType<typeof useAppDispatch>;

describe('Countdown Form Add View', () => {
  it('matches snapshot', async () => {
    const component = render(
      <CountdownFormAddView />
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  describe('Countdown Form - Handlers', () => {

    beforeEach(() => {
      mockAppDispatch = jest.fn();
      (useAppDispatch as jest.Mock).mockReturnValue(mockAppDispatch);
    });

    it('triggers onSubmit when clicking on the submit button', async () => {
      const component = render(
        <AppStateProvider>
          <CountdownFormAddView />
        </AppStateProvider>
      );

      // Get Go Back Element
      const goBackElement = await component.findByTestId('go-back');
      expect(goBackElement).toBeInTheDocument();

      // Trigger Click
      fireEvent.click(goBackElement);

      await waitFor(() => {
        expect(mockAppDispatch).toHaveBeenCalledWith({
          activeView: null,
          type: 'setActiveView'
        });
      });
    });
  });
});