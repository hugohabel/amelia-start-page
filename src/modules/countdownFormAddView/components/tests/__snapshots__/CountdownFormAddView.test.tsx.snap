// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown Form Add View matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="backContainer"
    data-testid="go-back"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
      />
    </svg>
    <span>
      Go Back
    </span>
  </div>
  <h2
    class="title"
  >
    New countdown
  </h2>
  <p
    class="subtitle"
  >
    Add the details for your countdown widget
  </p>
  <form
    autocomplete="off"
  >
    <input
      id="eventName"
      name="eventName"
      placeholder="Event name"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
